manager/manager.go

```diff
@@ -37,105 +37,167 @@ import (
+	var url string
 
 	for _, item := range items {
 		id := item.Identifier
 		norn, err := model.ConvertItemToNorn(item)
 		if err != nil {
 			log.WithPrefix(m.logger, level.Key(), level.ErrorValue()).Log(logging.MessageKey(), "failed to convert Item to Norn", "item", item)
 		}
-		if _, ok := m.nornsDispatch[id]; ok {
-			recentMap[id] = norn
+		if (norn.Destination.AWSConfig) == (model.AWSConfig{}) {
+			url = norn.Destination.HttpConfig.URL
 		} else {
-			dispatcher, err := dispatch.NewDispatcher(m.dispatcherConfig, norn, transport)
-			if err == nil {
-				newNorns = append(newNorns, nornDispatcher{norn: norn, dispatcher: dispatcher})
-				recentMap[id] = norn
-			} else {
+			url = norn.Destination.AWSConfig.Sqs.QueueURL
+		}
+
+		if _, ok := m.idFilter[id]; ok {
+			recentIDMap[id] = norn
+		} else {
+			filter, err = dispatch.NewFilter(m.filterConfig, dispatcher, norn, m.sender)
+			if err != nil {
 				m.logger.Log(level.Key(), level.ErrorValue(), logging.MessageKey(), err.Error)
+			} else {
+				recentIDMap[id] = norn
 			}
+
+		}
+
+		if _, ok := m.urlDispatcher[url]; ok {
+			recentURLMap[url] = norn
+		} else {
+			if (norn.Destination.AWSConfig) == (model.AWSConfig{}) {
+				dispatcher, err = dispatch.NewHttpDispatcher(m.dispatcherConfig, norn.Destination.HttpConfig, m.sender, m.logger, m.measures)
+				if err != nil {
+					m.logger.Log(level.Key(), level.ErrorValue(), logging.MessageKey(), err.Error)
+				} else {
+					newNorns = append(newNorns, filterDispatcher{norn: norn, dispatcher: dispatcher, filter: filter})
+					recentURLMap[url] = norn
+				}
+			} else {
+				dispatcher, err = dispatch.NewSqsDispatcher(m.dispatcherConfig, norn.Destination.AWSConfig, m.logger, m.measures)
+				if err != nil {
+					m.logger.Log(level.Key(), level.ErrorValue(), logging.MessageKey(), err.Error)
+				} else {
+					newNorns = append(newNorns, filterDispatcher{norn: norn, dispatcher: dispatcher, filter: filter})
+					recentURLMap[url] = norn
+				}
+			}
+		}
+	}
+
+	for i, filterdis := range m.idFilter {
+		if norn, ok := recentIDMap[i]; !ok {
+			oldNorns = append(oldNorns, filterdis.filter)
+			delete(m.idFilter, i)
+		} else {
+			filterdis.filter.Update(norn)
 		}
 	}
 
-	for i, norndis := range m.nornsDispatch {
-		if val, ok := recentMap[i]; !ok {
-			oldNorns = append(oldNorns, norndis.dispatcher)
-			delete(m.nornsDispatch, i)
+	for i, filterdis := range m.urlDispatcher {
+		if norn, ok := recentURLMap[i]; !ok {
+			oldNorns = append(oldNorns, filterdis.filter)
+			delete(m.urlDispatcher, i)
 		} else {
-			norndis.dispatcher.Update(val)
+			filterdis.dispatcher.Update(norn)
 		}
 	}
 
-	for _, norndis := range newNorns {
+	for _, filterdis := range newNorns {
 		m.mutex.Lock()
-		m.nornsDispatch[norndis.norn.DeviceID] = norndis
+		m.idFilter[filterdis.norn.DeviceID] = filterdis
 		m.mutex.Unlock()
 	}
 
-	for _, dispatcher := range oldNorns {
-		dispatcher.Stop(nil)
+	for _, filter := range oldNorns {
+		filter.Stop(nil)
 	}
```

Comment:
We should be adding and removing dispatchers in this same way - within a lock and at the end, once we've sorted through everything.